### PR TITLE
Add DeepONet implemented by Pytorch

### PR DESCRIPTION
**Pull request by Jerry-Jzy**
## what is the purpose of this change:
Add DeepONet implemented by Pytorch

### CODE REVIEW

**Comment by lululxvi** [comment]
Remove those lines related to learning rate first.

**Comment by Jerry-Jzy** [comment]
> Remove those lines related to learning rate first.

Fixed

**Comment by lululxvi** [comment]
Not fixed. I still can see them.

**Comment by Jerry-Jzy** [comment]
> Not fixed. I still can see them.

How about the latest version?

File: deepxde/model.py [hunk 1]
```diff
@@ -412,7 +421,7 @@ def _outputs(self, training, inputs):
         if backend_name == "tensorflow.compat.v1":
             feed_dict = self.net.feed_dict(training, inputs)
             return self.sess.run(self.outputs, feed_dict=feed_dict)
-        if backend_name in ["tensorflow", "pytorch", "paddle"]:
+        if backend_name in ["tensorflow", "pytorch"]:
```
Comment (lululxvi) [inline deepxde/model.py:424]: Why do you delete "paddle"?

File: deepxde/model.py [hunk 2]
```diff
@@ -246,8 +246,17 @@ def outputs(training, inputs):
 
         def outputs_losses(training, inputs, targets, losses_fn):
             self.net.train(mode=training)
-            inputs = torch.as_tensor(inputs)
-            inputs.requires_grad_()
+            if isinstance(inputs, tuple):
+                multi_inputs = [torch.as_tensor(inputs[0])]
```
Comment (lululxvi) [inline deepxde/model.py:250]: Why `multi_inputs[0]` doesn't call `requires_grad_()`?

File: deepxde/nn/pytorch/deeponet.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+        if callable(layer_sizes_branch[1]):
+            # User-defined network
+            self.branch = layer_sizes_branch[1]
+        else:
+            # Fully connected network
+            self.branch = FNN(layer_sizes_branch, activation_branch, kernel_initializer)
+        self.trunk = FNN(layer_sizes_trunk, self.activation_trunk, kernel_initializer)
+        self.b = torch.tensor(0.0, requires_grad=True)
+
+    def forward(self, inputs, training=False):
```
Comment (lululxvi) [inline deepxde/nn/pytorch/deeponet.py:46]:
```python
def forward(self, inputs):
```

File: deepxde/nn/pytorch/deeponet.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+            # Fully connected network
+            self.branch = FNN(layer_sizes_branch, activation_branch, kernel_initializer)
+        self.trunk = None
+        if layer_sizes_trunk is not None:
+            self.trunk = FNN(
+                layer_sizes_trunk, self.activation_trunk, kernel_initializer
+            )
+            self.b = torch.tensor(0.0, requires_grad=True)
+
+    def forward(self, inputs, training=False):
```
Comment (lululxvi) [inline deepxde/nn/pytorch/deeponet.py:125]:
```
def forward(self, inputs):
```

File: deepxde/nn/pytorch/deeponet.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        if isinstance(activation, dict):
+            activation_branch = activation["branch"]
+            self.activation_trunk = activations.get(activation["trunk"])
+        else:
+            activation_branch = self.activation_trunk = activations.get(activation)
+
+        self.pod_basis = torch.as_tensor(
+            pod_basis, dtype=torch.float32
+        )
+        self.pod_basis.requires_grad_()
```
Comment (lululxvi) [inline deepxde/nn/pytorch/deeponet.py:111]: No `requires_grad_()`

File: deepxde/nn/pytorch/deeponet.py [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        layer_sizes_trunk=None,
+    ):
+        super().__init__()
+        if isinstance(activation, dict):
+            activation_branch = activation["branch"]
+            self.activation_trunk = activations.get(activation["trunk"])
+        else:
+            activation_branch = self.activation_trunk = activations.get(activation)
+
+        self.pod_basis = torch.as_tensor(
```
Comment (lululxvi) [inline deepxde/nn/pytorch/deeponet.py:108]: Reformat via black.

File: deepxde/model.py (hunk 1 above)
Comment (Jerry-Jzy) [reply deepxde/model.py]: has recovered it. but I don't remember modifying it. it is weird

File: deepxde/model.py (hunk 2 above)
Comment (Jerry-Jzy) [reply deepxde/model.py]: Fixed

File: deepxde/nn/pytorch/deeponet.py (hunk 3 above)
Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/deeponet.py]: Fixed

File: deepxde/nn/pytorch/deeponet.py (hunk 4 above)
Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/deeponet.py]: Fixed

File: deepxde/nn/pytorch/deeponet.py (hunk 5 above)
Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/deeponet.py]: Fixed

File: deepxde/nn/pytorch/deeponet.py (hunk 6 above)
Comment (Jerry-Jzy) [reply deepxde/nn/pytorch/deeponet.py]: Fixed

File: deepxde/model.py [hunk 7]
```diff
@@ -246,8 +246,17 @@ def outputs(training, inputs):
...
-            inputs = torch.as_tensor(inputs)
-            inputs.requires_grad_()
+            if isinstance(inputs, tuple):
+                multi_inputs = [torch.as_tensor(inputs[0]).requires_grad_()]
+                inputs_num = len(inputs)
+                for i in range(1, inputs_num):
+                    inputs_i = torch.as_tensor(inputs[i])
+                    inputs_i.requires_grad_()
+                    multi_inputs.append(inputs_i)
+                inputs = multi_inputs
```
Comment (lululxvi) [inline deepxde/model.py:256]:
```python
inputs = tuple(map(lambda x: torch.as_tensor(x).requires_grad_(), inputs))
```

Comment (Jerry-Jzy) [reply deepxde/model.py]: Fixed

File: deepxde/model.py [hunk 8]
```diff
@@ -246,8 +246,11 @@ def outputs(training, inputs):
 
         def outputs_losses(training, inputs, targets, losses_fn):
             self.net.train(mode=training)
-            inputs = torch.as_tensor(inputs)
-            inputs.requires_grad_()
+            if isinstance(inputs, tuple):
```
Comment (lululxvi) [inline deepxde/model.py:249]: You should also modify line 245.

Comment (Jerry-Jzy) [reply deepxde/model.py:253]: Fixed

File: deepxde/model.py [hunk 9]
```diff
@@ -242,7 +242,11 @@ def _compile_pytorch(self, lr, loss_fn, decay, loss_weights):
         def outputs(training, inputs):
             self.net.train(mode=training)
             with torch.no_grad():
-                return self.net(torch.as_tensor(inputs))
+                if isinstance(inputs, tuple):
+                    inputs = tuple(map(lambda x: torch.as_tensor(x), inputs))
```
Comment (lululxvi) [inline deepxde/model.py:246]:
```python
inputs = tuple(map(torch.as_tensor, inputs))
```

Comment (Jerry-Jzy) [reply deepxde/model.py]: Fixed
